File: core/src/com/ore/infinium/HotbarInventoryView.java
package com.ore.infinium;

import com.artemis.ComponentMapper;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.scenes.scene2d.*;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.scenes.scene2d.utils.DragAndDrop;
import com.badlogic.gdx.scenes.scene2d.utils.TextureRegionDrawable;
import com.badlogic.gdx.utils.Scaling;
import com.ore.infinium.components.BlockComponent;
import com.ore.infinium.components.ItemComponent;
import com.ore.infinium.components.SpriteComponent;
import com.ore.infinium.systems.NetworkClientSystem;
import com.ore.infinium.systems.TileRenderSystem;

/**
 * ***************************************************************************
 * Copyright (C) 2015 by Shaun Reich <sreich02@gmail.com>                *
 * *
 * This program is free software; you can redistribute it and/or            *
 * modify it under the terms of the GNU General Public License as           *
 * published by the Free Software Foundation; either version 2 of           *
 * the License, or (at your option) any later version.                      *
 * *
 * This program is distributed in the hope that it will be useful,          *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of           *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            *
 * GNU General Public License for more details.                             *
 * *
 * You should have received a copy of the GNU General Public License        *
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.    *
 * ***************************************************************************
 */
public class HotbarInventoryView implements Inventory.SlotListener {
    private Skin m_skin;
    private Table container;
    private SlotElement[] m_slots = new SlotElement[Inventory.maxHotbarSlots];
    private OreWorld m_world;

    private ComponentMapper<ItemComponent> itemMapper;
    private ComponentMapper<BlockComponent> blockMapper;
    private ComponentMapper<SpriteComponent> spriteMapper;

    //the model for this view
    private Inventory m_hotbarInventory;

    //the main player inventory, for drag and drop
    private Inventory m_inventory;

    private Label m_tooltip;
    private Stage m_stage;

    public HotbarInventoryView(Stage stage, Skin skin, Inventory hotbarInventory, Inventory inventory,
                               DragAndDrop dragAndDrop, OreWorld world) {
        m_skin = skin;
        m_inventory = inventory;
        m_world = world;
        m_stage = stage;

        m_world.m_artemisWorld.inject(this);

        m_hotbarInventory = hotbarInventory;
        //attach to the inventory model
        m_hotbarInventory.addListener(this);

        container = new Table(m_skin);
        container.setFillParent(true);
        container.top().left().setSize(800, 100);
        container.padLeft(10).padTop(10);

        container.defaults().space(4);

        stage.addActor(container);

        Image dragImage = new Image();
        dragImage.setSize(32, 32);

        for (byte i = 0; i < Inventory.maxHotbarSlots; ++i) {

            Image slotImage = new Image();

            SlotElement element = new SlotElement();
            m_slots[i] = element;

            element.itemImage = slotImage;

            Table slotTable = new Table(m_skin);
            element.table = slotTable;
            slotTable.setTouchable(Touchable.enabled);
            slotTable.addListener(new SlotClickListener(this, i));
            slotTable.addListener(new SlotInputListener(this, i));

            slotTable.add(slotImage);
            slotTable.background("default-pane");

            slotTable.row();

            Label itemCount = new Label(null, m_skin);
            slotTable.add(itemCount).bottom().fill();
            element.itemCountLabel = itemCount;

            //            container.add(slotTable).size(50, 50);
            container.add(slotTable).fill().size(50, 50);
            setHotbarSlotVisible(i, false);

            dragAndDrop.addSource(new HotbarDragSource(slotTable, i, dragImage, this));

            dragAndDrop.addTarget(new HotbarDragTarget(slotTable, i, this));
        }

        m_tooltip = new Label(null, m_skin);
        stage.addActor(m_tooltip);
    }

    private void deselectPreviousSlot() {
        m_slots[m_hotbarInventory.m_previousSelectedSlot].table.setColor(Color.WHITE);
    }

    @Override
    public void countChanged(byte index, Inventory inventory) {
        ItemComponent itemComponent = itemMapper.get(inventory.itemEntity(index));
        m_slots[index].itemCountLabel.setText(Integer.toString(itemComponent.stackSize));
    }

    @Override
    public void set(byte index, Inventory inventory) {
        SlotElement slot = m_slots[index];

        int itemEntity = inventory.itemEntity(index);
        ItemComponent itemComponent = itemMapper.get(itemEntity);
        m_slots[index].itemCountLabel.setText(Integer.toString(itemComponent.stackSize));

        TextureRegion region;
        SpriteComponent spriteComponent = spriteMapper.get(itemEntity);
        if (blockMapper.get(itemEntity) != null) {
            //fixme this concat is pretty...iffy
            region = m_world.m_artemisWorld.getSystem(TileRenderSystem.class).m_tilesAtlas.findRegion(
                    spriteComponent.textureName.concat("-00"));
        } else {
            region = m_world.m_atlas.findRegion(spriteComponent.textureName);
        }

        assert region != null : "textureregion for inventory item was not found!";

        Image slotImage = slot.itemImage;
        //        //m_blockAtlas.findRegion("stone"));

        slotImage.setDrawable(new TextureRegionDrawable(region));
        slotImage.setSize(region.getRegionWidth(), region.getRegionHeight());
        slotImage.setScaling(Scaling.fit);

        setHotbarSlotVisible(index, true);

        //do not exceed the max size/resort to horrible upscaling. prefer native size of each inventory sprite.
        //.maxSize(region.getRegionWidth(), region.getRegionHeight()).expand().center();
    }

    @Override
    public void removed(byte index, Inventory inventory) {
        SlotElement slot = m_slots[index];
        //       slot.itemImage.setDrawable(null);
        //        slot.itemCountLabel.setText(null);
        setHotbarSlotVisible(index, false);
    }

    @Override
    public void selected(byte index, Inventory inventory) {
        deselectPreviousSlot();
        m_slots[index].table.setColor(0, 0, 1, 1);
    }

    //FIXME: do the same for InventoryView
    private void setHotbarSlotVisible(int index, boolean visible) {
        if (!visible) {
            m_slots[index].itemImage.setDrawable(null);
            m_slots[index].itemCountLabel.setText(null);
        }
        m_slots[index].itemCountLabel.setVisible(visible);
        m_slots[index].itemImage.setVisible(visible);
    }

    private static class HotbarDragSource extends DragAndDrop.Source {
        private final byte index;
        private Image dragImage;
        private HotbarInventoryView hotbarInventoryView;

        public HotbarDragSource(Table slotTable, byte index, Image dragImage, HotbarInventoryView hotbarInventoryView) {
            super(slotTable);
            this.index = index;
            this.dragImage = dragImage;
            this.hotbarInventoryView = hotbarInventoryView;
        }

        public DragAndDrop.Payload dragStart(InputEvent event, float x, float y, int pointer) {
            //invalid drag start, ignore.
            if (hotbarInventoryView.m_hotbarInventory.itemEntity(index) == OreWorld.ENTITY_INVALID) {
                return null;
            }

            DragAndDrop.Payload payload = new DragAndDrop.Payload();

            InventorySlotDragWrapper dragWrapper = new InventorySlotDragWrapper();
            payload.setObject(dragWrapper);
            dragWrapper.type = Inventory.InventoryType.Hotbar;
            dragWrapper.dragSourceIndex = index;

            payload.setDragActor(dragImage);
            payload.setValidDragActor(dragImage);
            payload.setInvalidDragActor(dragImage);

            return payload;
        }
    }

    private static class HotbarDragTarget extends DragAndDrop.Target {
        private final byte index;
        private HotbarInventoryView inventory;

        public HotbarDragTarget(Table slotTable, byte index, HotbarInventoryView inventory) {
            super(slotTable);
            this.index = index;
            this.inventory = inventory;
        }

        public boolean drag(DragAndDrop.Source source, DragAndDrop.Payload payload, float x, float y, int pointer) {
            if (payload == null) {
                return false;
            }

            if (isValidDrop(payload)) {
                getActor().setColor(Color.GREEN);
                payload.getDragActor().setColor(0, 1, 0, 1);

                return true;
            } else {
                getActor().setColor(Color.RED);
                payload.getDragActor().setColor(1, 0, 0, 1);
            }

            return false;
        }

        private boolean isValidDrop(DragAndDrop.Payload payload) {

            InventorySlotDragWrapper dragWrapper = (InventorySlotDragWrapper) payload.getObject();
            if (dragWrapper.dragSourceIndex != index) {
                //maybe make it green? the source/dest is not the same
                if (inventory.m_hotbarInventory.itemEntity(index) == OreWorld.ENTITY_INVALID) {
                    //only make it green if the slot is empty
                    return true;
                }
            }

            return false;
        }

        public void reset(DragAndDrop.Source source, DragAndDrop.Payload payload) {
            payload.getDragActor().setColor(1, 1, 1, 1);
            getActor().setColor(Color.WHITE);
            //restore selection, it was just dropped..
            inventory.selected(inventory.m_hotbarInventory.m_selectedSlot, inventory.m_hotbarInventory);
        }

        public void drop(DragAndDrop.Source source, DragAndDrop.Payload payload, float x, float y, int pointer) {
            InventorySlotDragWrapper dragWrapper = (InventorySlotDragWrapper) payload.getObject();

            //ensure the dest is empty before attempting any drag & drop!
            if (inventory.m_hotbarInventory.itemEntity(this.index) != OreWorld.ENTITY_INVALID) {
                return;
            }

            if (dragWrapper.type == Inventory.InventoryType.Hotbar) {
                //move the item from the source to the dest (from hotbarinventory to hotbarinventory)
                inventory.m_hotbarInventory.setSlot(this.index, inventory.m_hotbarInventory.itemEntity(
                        dragWrapper.dragSourceIndex));
                inventory.m_world.m_artemisWorld.getSystem(NetworkClientSystem.class)
                                                .sendInventoryMove(Inventory.InventoryType.Hotbar,
                                                                   dragWrapper.dragSourceIndex,
                                                                   Inventory.InventoryType.Hotbar, index);

                //remove the source item
                inventory.m_hotbarInventory.takeItem(dragWrapper.dragSourceIndex);
            } else {
                //main inventory

                //move the item from the source to the dest (from main inventory, to this hotbar inventory)
                inventory.m_hotbarInventory.setSlot(this.index,
                                                    inventory.m_inventory.itemEntity(dragWrapper.dragSourceIndex));
                //fixme?                    inventory.m_previousSelectedSlot = index;
                inventory.m_world.m_artemisWorld.getSystem(NetworkClientSystem.class)
                                                .sendInventoryMove(Inventory.InventoryType.Inventory,
                                                                   dragWrapper.dragSourceIndex,
                                                                   Inventory.InventoryType.Hotbar, index);

                //remove the source item
                inventory.m_inventory.takeItem(dragWrapper.dragSourceIndex);
            }

            //select new index
            inventory.m_hotbarInventory.selectSlot(index);
        }

    }

    private static class SlotInputListener extends InputListener {
        private byte index;
        private HotbarInventoryView inventory;

        SlotInputListener(HotbarInventoryView inventory, byte index) {
            this.index = index;
            this.inventory = inventory;
        }

        @Override
        public boolean mouseMoved(InputEvent event, float x, float y) {
            int itemEntity = inventory.m_hotbarInventory.itemEntity(index);
            if (itemEntity != OreWorld.ENTITY_INVALID) {
                inventory.m_tooltip.setVisible(true);

                inventory.m_tooltip.setPosition(Gdx.input.getX(), Gdx.graphics.getHeight() - Gdx.input.getY() - 50);

                //fixme, obviously texture name is not a valid tooltip text. we need a real name, but should it be in
                // sprite or item? everything should probably have a canonical name, no?
                ItemComponent itemComponent = inventory.itemMapper.get(itemEntity);
                SpriteComponent spriteComponent = inventory.spriteMapper.get(itemEntity);
                inventory.m_tooltip.setText(spriteComponent.textureName);
            } else {
                inventory.m_tooltip.setVisible(false);
            }

            return super.mouseMoved(event, x, y);
        }

        @Override
        public void exit(InputEvent event, float x, float y, int pointer, Actor toActor) {

            super.exit(event, x, y, pointer, toActor);
        }
    }

    private static class SlotClickListener extends ClickListener {
        private byte index;
        private HotbarInventoryView inventory;

        public SlotClickListener(HotbarInventoryView inventory, byte index) {
            this.inventory = inventory;
            this.index = index;
        }

        @Override
        public void clicked(InputEvent event, float x, float y) {
            inventory.deselectPreviousSlot();
            inventory.m_hotbarInventory.selectSlot(index);
        }
    }

    private class SlotElement {
        public Image itemImage;
        public Label itemCountLabel;
        public Table table;
    }
}

File: core/src/com/ore/infinium/HotbarInventoryView.kt
/**
MIT License

Copyright (c) 2016 Shaun Reich <sreich02@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
 */

package com.ore.infinium

import com.artemis.ComponentMapper
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.Touchable
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.scenes.scene2d.ui.Tooltip
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.scenes.scene2d.utils.DragAndDrop
import com.badlogic.gdx.scenes.scene2d.utils.TextureRegionDrawable
import com.badlogic.gdx.utils.Scaling
import com.kotcrab.vis.ui.VisUI
import com.kotcrab.vis.ui.widget.Tooltip.TooltipStyle
import com.kotcrab.vis.ui.widget.VisImage
import com.kotcrab.vis.ui.widget.VisLabel
import com.kotcrab.vis.ui.widget.VisTable
import com.ore.infinium.components.BlockComponent
import com.ore.infinium.components.ItemComponent
import com.ore.infinium.components.SpriteComponent
import com.ore.infinium.systems.client.ClientNetworkSystem
import com.ore.infinium.systems.client.MultiRenderSystem
import com.ore.infinium.util.*

class HotbarInventoryView(private val stage: Stage,
        //the model for this view
                          private val inventory: HotbarInventory,
        //the main player inventory, for drag and drop
                          dragAndDrop: DragAndDrop,
                          private val world: OreWorld) : Inventory.SlotListener {

    private val container: VisTable
    private val slots = mutableListOf<SlotElement>()

    //    private lateinit var tileRenderSystem: TileRenderSystem
    private lateinit var multiRenderSystem: MultiRenderSystem

    private lateinit var mItem: ComponentMapper<ItemComponent>
    private lateinit var mBlock: ComponentMapper<BlockComponent>
    private lateinit var mSprite: ComponentMapper<SpriteComponent>

    private val tooltip: Tooltip<VisTable>
    private val tooltipLabel: VisLabel

    init {
        world.artemisWorld.inject(this)

        //attach to the inventory model
        inventory.addListener(this)

        container = VisTable().apply {
            setFillParent(true)
            top().left().setSize(800f, 100f)
            padLeft(10f).padTop(10f)
            defaults().space(4f)
        }

        stage.addActor(container)

        val dragImage = VisImage()
        dragImage.setSize(32f, 32f)

        repeat(Inventory.maxHotbarSlots) {
            val element = SlotElement(this, it)
            slots.add(element)
            container.add(element.slotTable).size(50f, 50f)

            setHotbarSlotVisible(it, false)

            dragAndDrop.addSource(HotbarDragSource(element.slotTable, it, dragImage, this))

            dragAndDrop.addTarget(HotbarDragTarget(element.slotTable, it, this))

        }

        val style = VisUI.getSkin().get("default", TooltipStyle::class.java)

        tooltipLabel = VisLabel()
        val tooltipTable = VisTable().apply {
            add(tooltipLabel)
            background = style.background
        }

        tooltip = Tooltip<VisTable>(tooltipTable)
    }

    private fun deselectPreviousSlot() {
        slots[inventory.previousSelectedSlot].slotTable.color = Color.WHITE
    }

    override fun slotItemCountChanged(index: Int, inventory: Inventory) {
        val cItem = mItem.get(inventory.itemEntity(index))
        slots[index].itemCount.setText(cItem.stackSize.toString())
    }

    override fun slotItemChanged(index: Int, inventory: Inventory) {
        val slot = slots[index]

        val itemEntity = inventory.itemEntity(index)
        if (isInvalidEntity(itemEntity)) {
            setHotbarSlotVisible(index, false)
            return
        }

        val cItem = mItem.get(itemEntity)
        slots[index].itemCount.setText(cItem.stackSize.toString())

        val cSprite = mSprite.get(itemEntity)

        val region = textureForInventoryItem(itemEntity, cSprite.textureName!!)

        val slotImage = slot.itemImage

        slotImage.drawable = TextureRegionDrawable(region)
        slotImage.setSize(region.regionWidth.toFloat(), region.regionHeight.toFloat())
        slotImage.setScaling(Scaling.fit)

        setHotbarSlotVisible(index, true)

        //do not exceed the max size/resort to horrible upscaling. prefer native size of each inventory sprite.
        //.maxSize(region.getRegionWidth(), region.getRegionHeight()).expand().center();
    }

    //fixme this is also duped in the InventoryView, but not sure where else to put it...the current way is a hack anyway
    fun textureForInventoryItem(itemEntity: Int, textureName: String): TextureRegion {
        var region: TextureRegion?
        if (mBlock.opt(itemEntity) != null) {
            //fixme this concat is pretty...iffy
            region = multiRenderSystem.tileRenderSystem.tilesAtlas.findRegion("$textureName-00")

            if (region == null) {
                //try again but without -00
                region = multiRenderSystem.tileRenderSystem.tilesAtlas.findRegion(textureName)
            }
        } else {
            region = world.atlas.findRegion(textureName)
        }

        assert(region != null) { "textureregion for inventory item entity id: $itemEntity, was not found!" }

        return region!!
    }

    override fun slotItemRemoved(index: Int, inventory: Inventory) {
        setHotbarSlotVisible(index, false)
    }

    override fun slotItemSelected(index: Int, inventory: Inventory) {
        deselectPreviousSlot()
        slots[index].slotTable.setColor(0f, 0f, 1f, 1f)
    }

    //FIXME: do the same for InventoryView
    /**
     * hides/shows the text and image for this index. For e.g.
     * when an item leaves, or init time
     */
    private fun setHotbarSlotVisible(index: Int, visible: Boolean) {
        if (!visible) {
            slots[index].itemImage.drawable = null
            slots[index].itemCount.setText(null)
        }
        slots[index].itemCount.isVisible = visible
        slots[index].itemImage.isVisible = visible
    }

    private class HotbarDragSource(slotTable: Table,
                                   private val index: Int,
                                   private val dragImage: Image,
                                   private val hotbarInventoryView: HotbarInventoryView) : DragAndDrop.Source(
            slotTable) {

        override fun dragStart(event: InputEvent, x: Float, y: Float, pointer: Int): DragAndDrop.Payload? {
            //invalid drag start, ignore.
            if (isInvalidEntity(hotbarInventoryView.inventory.itemEntity(index))) {
                return null
            }

            val payload = DragAndDrop.Payload()

            val dragWrapper = InventorySlotDragWrapper(sourceInventory = hotbarInventoryView.inventory,
                                                       dragSourceIndex = index)
            payload.`object` = dragWrapper

            payload.dragActor = dragImage
            payload.validDragActor = dragImage
            payload.invalidDragActor = dragImage

            return payload
        }
    }

    private class HotbarDragTarget(slotTable: VisTable,
                                   private val index: Int,
                                   private val inventoryView: HotbarInventoryView) : DragAndDrop.Target(
            slotTable) {

        override fun drag(source: DragAndDrop.Source,
                          payload: DragAndDrop.Payload?,
                          x: Float,
                          y: Float,
                          pointer: Int): Boolean {
            payload ?: return false

            if (isValidDrop(payload)) {
                BaseInventoryView.setSlotColor(payload, actor, Color.GREEN)
                return true
            } else {
                BaseInventoryView.setSlotColor(payload, actor, Color.RED)
                //reject
                return false
            }
        }

        private fun isValidDrop(payload: DragAndDrop.Payload): Boolean {
            val dragWrapper = payload.`object` as InventorySlotDragWrapper
            if (dragWrapper.dragSourceIndex == index &&
                    dragWrapper.sourceInventory.inventoryType == inventoryView.inventory.inventoryType) {
                //trying to drop on the same slot, on the same inventory
                return false
            }

            if (isInvalidEntity(inventoryView.inventory.itemEntity(index))) {
                //only make it green if the slot is empty
                return true
            }

            return false
        }

        override fun reset(source: DragAndDrop.Source?, payload: DragAndDrop.Payload?) {
            payload ?: error("payload is null. bad state")

            BaseInventoryView.setSlotColor(payload, actor, Color.WHITE)

            //restore selection, it was just dropped..
            inventoryView.slotItemSelected(inventoryView.inventory.selectedSlot, inventoryView.inventory)
        }

        override fun drop(source: DragAndDrop.Source, payload: DragAndDrop.Payload, x: Float, y: Float, pointer: Int) {
            val dragWrapper = payload.`object` as InventorySlotDragWrapper
            val hotbarInventory = inventoryView.inventory

            //ensure the dest is empty before attempting any drag & drop!
            if (!isValidDrop(payload)) {
                return
            }


            val itemEntity = dragWrapper.sourceInventory.itemEntity(dragWrapper.dragSourceIndex)

            //grab item from source, copy it into dest
            hotbarInventory.setSlot(this.index, itemEntity)

            //fixme?                    inventory.previousSelectedSlot = index;

            val clientNetworkSystem = inventoryView.world.artemisWorld.system<ClientNetworkSystem>()
            clientNetworkSystem.sendInventoryMove(sourceInventoryType = dragWrapper.sourceInventory.inventoryType,
                                                  sourceIndex = dragWrapper.dragSourceIndex,
                                                  destInventoryType = inventoryView.inventory.inventoryType,
                                                  destIndex = index)

            //remove the source item now that the move is complete
            dragWrapper.sourceInventory.takeItem(dragWrapper.dragSourceIndex)


            //select new index
            hotbarInventory.selectSlot(index)
        }
    }

    private class test : OreInputListener() {
    }

    private class SlotInputListener internal constructor(private val inventory: HotbarInventoryView,
                                                         private val index: Int)
        : OreInputListener() {
        override fun enter(event: InputEvent, x: Float, y: Float, pointer: Int, fromActor: Actor?) {
            val itemEntity = inventory.inventory.itemEntity(index)
            if (isValidEntity(itemEntity)) {
                inventory.tooltip.enter(event, x, y, pointer, fromActor)
            }

            super.enter(event, x, y, pointer, fromActor)
        }

        override fun mouseMoved(event: InputEvent, x: Float, y: Float): Boolean {
            inventory.tooltip.mouseMoved(event, x, y)

            val itemEntity = inventory.inventory.itemEntity(index)

            if (isValidEntity(itemEntity)) {
                val cItem = inventory.mItem.get(itemEntity)
                val cSprite = inventory.mSprite.get(itemEntity)
                inventory.tooltipLabel.setText(cItem.name)
            }

            return super.mouseMoved(event, x, y)
        }

        override fun exit(event: InputEvent, x: Float, y: Float, pointer: Int, toActor: Actor?) {
            inventory.tooltip.exit(event, x, y, pointer, toActor)

            super.exit(event, x, y, pointer, toActor)
        }
    }

    private class SlotClickListener(private val inventory: HotbarInventoryView,
                                    private val index: Int) : ClickListener() {

        override fun clicked(event: InputEvent?, x: Float, y: Float) {
            inventory.deselectPreviousSlot()
            inventory.inventory.selectSlot(index)
        }
    }

    private inner class SlotElement(inventoryView: HotbarInventoryView, index: Int) {
        val itemImage = VisImage()
        val slotTable = VisTable()
        val itemCount = VisLabel()

        init {
            with(slotTable) {
                touchable = Touchable.enabled
                add(itemImage)

                addListener(SlotClickListener(inventoryView, index))
                addListener(SlotInputListener(inventoryView, index))

                background("default-pane")

                row()

                add(itemCount).bottom().fill()
            }
        }
    }
}
